neural_speed/core/layers/mha_dense.cpp

```diff
@@ -85,14 +92,34 @@ struct attn_fwd_args_t {
 struct mha_problem_t {
   int batch_size, head_num, heads_kv, head_size, sl_q, sl_kv;
 };
-
-inline __m512 poly_scale_2nd_ps(const __m512 z, const __m512 f, const __m512 c0, const __m512 c1, const __m512 c2) {
+template <typename X_T, typename Z_T = X_T>
+inline X_T poly_scale_2nd_ps(const Z_T z, const X_T f, const X_T c0, const X_T c1, const X_T c2);
+template <>
+TARGET_512 inline __m512 poly_scale_2nd_ps<__m512, __m512>(const __m512 z, const __m512 f, const __m512 c0,
+                                                           const __m512 c1, const __m512 c2) {
   const auto y = _mm512_fmadd_ps(_mm512_fmadd_ps(f, c0, c1), f, c2);  // auto y = (f * c0 + c1) * f + c2;
   const auto exp = _mm512_scalef_ps(y, z);
   return exp;
 }
+template <>
+inline __m256 poly_scale_2nd_ps<__m256, __m256i>(const __m256i z, const __m256 f, const __m256 c0, const __m256 c1,
+                                                 const __m256 c2) {
+  const auto y = _mm256_fmadd_ps(_mm256_fmadd_ps(f, c0, c1), f, c2);  // auto y = (f * c0 + c1) * f + c2;
+  static const auto mask_exp = _mm256_set1_epi32(0x7f800000);
+  static const auto mask_not_exp = _mm256_set1_epi32(~0x7f800000);
+
+  const auto y_exp = _mm256_and_si256(_mm256_castps_si256(y), mask_exp);
+  const auto y_not_exp = _mm256_and_si256(_mm256_castps_si256(y), mask_not_exp);
+
+  const auto y_exp_scaled = _mm256_add_epi32(y_exp, _mm256_slli_epi32(z, 23));
+  return _mm256_castsi256_ps(_mm256_or_si256(y_not_exp, _mm256_and_si256(y_exp_scaled, mask_exp)));
+}
+
+template <typename X_T>
+inline X_T exp_ps_0_1(const X_T x);
 
-inline __m512 exp_ps_0_1(const __m512 x) {
+template <>
+TARGET_512 inline __m512 exp_ps_0_1<__m512>(const __m512 x) {
```

Comment:
how about moving them to bestla::kernel? And add kernel wrapper for them

Reply:
Moved in 480d287